### PR TITLE
YJIT: Allow testing assembler with disasm

### PR DESCRIPTION
**Pull request by k0kubun**
This PR proposes an alternative way to write assembler tests, rewriting tests of https://github.com/ruby/ruby/pull/7453 as a starter.

I don't really memorize how x86_64 instructions are encoded, so it's hard for me to read these test cases. Ideally, I'd like to test the generated code using disasm, assuming libcapstone is not broken. A challenge in it is that addresses in generated code are always different, so those values are replaced with `start_addr`-origin values.

### CODE REVIEW

**Comment by k0kubun** [comment]
How about this? https://github.com/ruby/ruby/pull/7470/commits/57a26a43db146c560449e1d1c285ee3f2a646a27 I removed new dependencies.

**Comment by maximecb** [comment]
Still not a huge fan of adding a dev dependency or relying on the output of capstone and string processing 😅 

Let's ask @XrXr what his opinion is when he is back next week.

**Comment by XrXr** [comment]
I thought about doing something like this before and never got around to it. Thanks for looking into this! Being able to see the disassembly make these much easier to read.

In terms of dev dependency, I think we can make it test the disassembly optionally. So we would have something like:
```
assert_disasm!(
    cb,
    "488d5b08", /* checked when cargo test is built without features=disasm */,
    "lea rbx, [rbx + 8]..." /* checked in `cargo test --all-features`, which we do on CI */
)
```
Some string manipulation to deindent is probably necessary, but maybe we can do less of it by not testing the addresses; I don't think the addresses are awfully important. I'm less concerned about Capstone because it seems to be reliable WRT always giving the same output.

**Review by maximecb — APPROVED** [review]

**Comment by maximecb** [comment]
I was on the fence, but since Alan is in favor, let's go forward with this 👍

**Comment by k0kubun** [comment]
> Some string manipulation to deindent is probably necessary, but maybe we can do less of it by not testing the addresses; I don't think the addresses are awfully important. I'm less concerned about Capstone because it seems to be reliable WRT always giving the same output.

Ah, I had forgotten that we were passing addresses to `disasm_all` ourselves. If we pass 0 to it, we wouldn't need address manipulation there. I'll fix it so.

**Comment by k0kubun** [comment]
addressed Alan's points in https://github.com/ruby/ruby/pull/7470/commits/ff6301b02b6cbf31dcf10b3e6fa8cad4e0149c97 and https://github.com/ruby/ruby/pull/7470/commits/05c9865e9cc8122f610c5cc3acc7848badcd194a.

File: yjit/src/disasm.rs [hunk 1]
```diff
@@ -195,6 +198,73 @@ pub fn disasm_addr_range(cb: &CodeBlock, start_addr: usize, end_addr: usize) ->
     return out;
 }
 
+/// Macro useful for raw string literals
```
Comment (maximecb) [inline yjit/src/disasm.rs:201]: Should explain in the comment what this macro does.

Comment (k0kubun) [reply yjit/src/disasm.rs]: Ah yes, I forgot updating this https://github.com/ruby/ruby/pull/7470/commits/86999bb77300dfb1bbc591db5e7d082bf6ff6ade. Thanks for noticing it.

**Pull request merged**